### PR TITLE
Properly handle categorization of glyph variants

### PR DESCRIPTION
**Pull request by madig**
See #386.

### CODE REVIEW

**Comment by madig** [comment]
Alright, this works. Now to clean it up.

@moyogo @belluzj comments?

File: tests/glyphdata_test.py [hunk 1]
```diff
@@ -32,6 +32,10 @@ def test_production_name(self):
         self.assertEqual(prod("s_t"), "s_t")
         self.assertEqual(prod("Gcommaaccent"), "uni0122")
         self.assertEqual(prod("o_f_f_i.foo"), "o_f_f_i.foo")
+        self.assertEqual(prod("brevecomb"), "uni0306")
+        self.assertEqual(prod("brevecomb.case"), "uni0306.case")
+        self.assertEqual(prod("brevecomb_acutecomb"), "u'uni03060301")
```
Comment (schriftgestalt) [inline tests/glyphdata_test.py]: > "u'uni03060301"
Is this correct?

**Comment by madig** [comment]
@schriftgestalt any more thoughts on this?

File: Lib/glyphsLib/glyphdata.py [hunk 2]
```diff
@@ -31,69 +31,224 @@
...
+
+    Note:
+    - Glyph name is the full name, e.g. "brevecomb_acutecomb.case".
+    - Base name is the base part, e.g. "brevecomb_acutecomb"
+    - Suffix is e.g. "case".
+    """
+
+    def is_unicode_u_value(name):
+        return name.startswith("u") and all(
+            part_char in "0123456789ABCDEF" for part_char in name[1:]
```
Comment (anthrotype) [inline Lib/glyphsLib/glyphdata.py]: use a set of chars instead of a string literal here maybe

File: Lib/glyphsLib/glyphdata.py [hunk 3]
```diff
@@ -31,69 +31,224 @@
...
+        or base_name in agl.AGL2UV
+        or glyph_name in (".notdef", ".null")
+    ):
+        return glyph_name
+
+    if glyph_name in data.PRODUCTION_NAMES:  # e.g. ain_alefMaksura-ar.fina -> uniFD13
+        return data.PRODUCTION_NAMES[glyph_name]
+    if base_name in data.PRODUCTION_NAMES:
+        final_production_name = data.PRODUCTION_NAMES[base_name] + dot + suffix
+        if len(final_production_name) > 31:
```
Comment (anthrotype) [inline Lib/glyphsLib/glyphdata.py]: are you sure about the length limit of 31? I remember it being 64, but maybe i'm wrong.

File: Lib/glyphsLib/glyphdata.py [hunk 4]
```diff
@@ -31,69 +31,224 @@
...
+        return None
+
+    # So we have a ligature that is not mapped in PRODUCTION_NAMES. Split it up and
+    # look up the individual parts.
+    base_name_parts = base_name.split("_")
+
+    # If all parts are in the AGLFN list, the glyph name is our production
+    # name already.
+    if all(part in agl.AGL2UV for part in base_name_parts):
+        if len(glyph_name) > 31:
```
Comment (anthrotype) [inline Lib/glyphsLib/glyphdata.py]: again, we should confirm the 31 length limit

File: Lib/glyphsLib/glyphdata.py [hunk 5]
```diff
@@ -31,69 +31,224 @@
...
+                    _character_outside_BMP = True
+
+            else:
+                return None
+
+    # Some names Glyphs uses resolve to other names that are not uniXXXX names and may
+    # contain dots (e.g. idotaccent -> i.loclTRK). If there is any name with a "." in
+    # it before the last element, punt. We'd have to introduce a "." into the ligature
+    # midway, which is invalid according to the AGL. Example: "a_i.loclTRK" is valid,
+    # but "a_i.loclTRK_a" isn't.
```
Comment (anthrotype) [inline Lib/glyphsLib/glyphdata.py]: that's correct according to AGLFN, I just hope that Glyphs.app's names also always conform to that rule.. Do they?

File: Lib/glyphsLib/glyphdata.py [hunk 6]
```diff
@@ -31,69 +31,224 @@
...
-
-
-def _get_unicode_category(unistr):
-    # We use data for a fixed Unicode version (3.2) so that our generated
-    # data files are independent of Python runtime that runs the rules.
-    # By switching to current Unicode data, we could save some entries
-    # in our exception tables, but the gains are not very large; only
-    # about one thousand entries.
-    if not unistr:
+        final_production_name = "{}".format("_".join(production_names)) + dot + suffix
```
Comment (anthrotype) [inline Lib/glyphsLib/glyphdata.py]: the `"{}".format(...)` here seems redundant, no? The `"_".join(production_names)` already returns a `str`.

File: Lib/glyphsLib/glyphdata.py [hunk 7]
```diff
@@ -31,69 +31,224 @@
...
+    thousand entries.
+    """
+
+    if not character:
+        return None
+
     if NARROW_PYTHON_BUILD:
-        utf32_str = unistr.encode("utf-32-be")
-        nchars = len(utf32_str)//4
+        utf32_str = character.encode("utf-32-be")
```
Comment (anthrotype) [inline Lib/glyphsLib/glyphdata.py]: why did you rename `unistr` argument to `character`? It's technically a string as it can contain multiple characters. The previous name made more sense I think.

File: tests/glyphdata_test.py [hunk 8]
```diff
@@ -44,6 +70,10 @@ def test_unicode(self):
         self.assertIsNone(uni("s_t"))  # no 'unicode' in GlyphsData
         self.assertEqual(uni("Gcommaaccent"), "Ģ")
         self.assertEqual(uni("o_f_f_i.foo"), "offi")
+        self.assertEqual(uni("brevecomb"), "̆")
```
Comment (anthrotype) [inline tests/glyphdata_test.py]: maybe use the `\u0123` escape instead of a literal unicode string here, as it's not really legible

File: Lib/glyphsLib/glyphdata.py (hunk 3 above)
Comment (madig) [reply Lib/glyphsLib/glyphdata.py]: https://github.com/adobe-type-tools/agl-specification

> Length and character set limitations on names Glyphs from Western OpenType/CFF and TrueType fonts must still be referenced in many workflows as name-keyed font data. As a result, glyph names are still subject to the length and character set limitations that are imposed by the Type 1 specification and PostScript interpreter implementations. Both of these require that a glyph name be no longer than 31 characters in length [...]

File: Lib/glyphsLib/glyphdata.py (hunk 5 above)
Comment (madig) [reply Lib/glyphsLib/glyphdata.py]: Glyphs turns "a_idotaccent_a" into "a_i.loclTRK_a".

File: Lib/glyphsLib/glyphdata.py (hunk 3 above)
Comment (anthrotype) [reply Lib/glyphsLib/glyphdata.py]: ok.. but the Adobe Feature File spec says 63...
https://github.com/adobe-type-tools/afdko/blob/develop/docs/OpenTypeFeatureFileSpecification.html

Comment (anthrotype) [reply Lib/glyphsLib/glyphdata.py]: I guess the 64th is a C-string NULL terminating byte :smile:

Comment (anthrotype) [reply Lib/glyphsLib/glyphdata.py]: I mean, if makeotf allows 64, who are we to restrict it to 31?!

Comment (madig) [reply Lib/glyphsLib/glyphdata.py]: Hmm... so, which one should it be? 😕 

@miguelsousa ?

**Comment by anthrotype** [comment]
some of this logic is generic and not-glyphs.app specific, it would be nice to integrate it in ufo2ft PostProcessor which creates the production names (when an explicit postscriptNames in fontinfo.plist isn't provided).
There is also this https://github.com/googlei18n/fontmake/issues/455

File: Lib/glyphsLib/glyphdata.py (hunk 3 above)
Comment (madig) [reply Lib/glyphsLib/glyphdata.py]: Ok, let's just assume everything is broken anyway and if makeotf is doing it, so should we 😁👍

File: Lib/glyphsLib/glyphdata.py (hunk 5 above)
Comment (anthrotype) [reply Lib/glyphsLib/glyphdata.py]: that's not good. What does your code do in that case?

File: Lib/glyphsLib/glyphdata.py (hunk 3 above)
Comment (anthrotype) [reply Lib/glyphsLib/glyphdata.py]: yes, let's use 63 and paste a link to that Feature File spec paragraph in an inline comment, thanks

**Comment by moyogo** [comment]
> are you sure about the length limit of 31? I remember it being 64, but maybe i'm wrong.

@anthrotype 

The [OpenType Feature File Specification]() does say.
> A glyph name may be up to 63 characters in length [...].

The [AGL-specification](https://github.com/adobe-type-tools/agl-specification) says:
> All glyph names a subject to a length-limit of 31 characters in some current software implementations, [...]. This length limit can be an issue with glyph names for ligature and ornaments.

File: Lib/glyphsLib/glyphdata.py (hunk 5 above)
Comment (madig) [reply Lib/glyphsLib/glyphdata.py]: https://github.com/googlei18n/glyphsLib/pull/387/files#diff-969a14671a2b5301f379a9ffe3bb13f6R56 😁

File: tests/glyphdata_test.py [hunk 9]
```diff
@@ -32,6 +33,31 @@ def test_production_name(self):
...
+        self.assertEqual(prod("Dboldscript-math_Dboldscript-math.f"), "u1D4D3_u1D4D3.f")
+        self.assertEqual(prod("Dboldscript-math_a"), "u1D4D3_a")
+        self.assertEqual(prod("a_Dboldscript-math"), "a_u1D4D3")  # a_Dboldscript-math
+        self.assertEqual(prod("Dboldscript-math_a_aa"), "u1D4D3_a_uniA733")  # Dboldscript-math_a_aa
+        self.assertEqual(prod("Dboldscript-math_a_aaa"), "Dboldscript-math_a_aaa")
+        self.assertEqual(prod("brevecomb_Dboldscript-math"), "uni0306_u1D4D3")  # brevecomb_Dboldscript-math
+        self.assertEqual(prod("Dboldscript-math_brevecomb"), "u1D4D3_uni0306")  # Dboldscript-math_brevecomb
+        self.assertEqual(prod("idotaccent"), "i.loclTRK")
+        self.assertEqual(prod("a_idotaccent"), "a_i.loclTRK")
+        self.assertEqual(prod("a_idotaccent_a"), "a_idotaccent_a")  # a_i.loclTRK_a
```
Comment (anthrotype) [inline tests/glyphdata_test.py]: makes sense

**Comment by moyogo** [comment]
For me, the OT fea spec defines what goes in the feature file, before anything is converted to production names. The production names that go in the post table need to follow the AGL spec.

**Comment by anthrotype** [comment]
anyway, it doesn't look like you're truncating them to 31. you just exit with None when it doesn't fit. What if the name that you didn't rename (because its production names exceeded 31) also exceeds the 31 char limit. That would also be invalid according to that aglfn spec

**Comment by madig** [comment]
Yes. I punt on that one and hope that someday, some tool is going to rename stuff that doesn't fit to glyphXXXXX ^_^ or just not write it to the `post` table.

File: Lib/glyphsLib/glyphdata.py (hunk 2 above)
Comment (madig) [reply Lib/glyphsLib/glyphdata.py]:
```
In [5]:  hex_characters = {
   ...:    "0", "1", "2", "3", "4", "5", "6", "7", "8", "9", "A", "B", "C", "D", "E", "F",
   ...:  }

In [10]: %timeit all(part_char in hex_characters for part_char in uname[1:])
623 ns ± 3.89 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [11]: %timeit all(part_char in "0123456789ABCDEF" for part_char in uname[1:])
567 ns ± 3.34 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```
Let's keep the string literal :)

**Comment by madig** [comment]
Regarding name length: maybe let's wait for clarification from Adobe.

File: Lib/glyphsLib/glyphdata.py (hunk 5 above)
Comment (schriftgestalt) [reply Lib/glyphsLib/glyphdata.py]: I check the ‘a_i.loclTRK_a’ name. That is clearly wrong.
